mfe.go

```diff
@@ -0,0 +1,1092 @@
+	enclosedFivePrimeIdx := closingFivePrimeIdx + 1
+
+	// seek to the first stem
+	for enclosedFivePrimeIdx <= closingThreePrimeIdx && pairTable[enclosedFivePrimeIdx] == -1 {
+		enclosedFivePrimeIdx++
+	}
+
+	// add inital unpaired nucleotides
+	nbUnpairedNucleotides := enclosedFivePrimeIdx - closingFivePrimeIdx - 1
+
+	// MLclosing is energetic penalty imposed when a base pair encloses a
+	// multi loop
+	energy := fc.energyParams.MLclosing
+	for enclosedFivePrimeIdx < closingThreePrimeIdx {
+		/* fivePrimeIter must have a pairing partner */
+		enclosedThreePrimeIdx := pairTable[enclosedFivePrimeIdx]
+		/* get type of the enclosed base pair (enclosedFivePrimeIdx,enclosedThreePrimeIdx) */
+		basePairType := encodedBasePairType(fc.sequence[enclosedFivePrimeIdx],
+			fc.sequence[enclosedThreePrimeIdx], fc.energyParams.basePairEncodedTypeMap)
+
+		enclosedFivePrimeMismatch := fc.encodedSequence[enclosedFivePrimeIdx-1]
+		enclosedThreePrimeMismatch := fc.encodedSequence[enclosedThreePrimeIdx+1]
+
+		energy += multiLoopStemEnergy(basePairType, enclosedFivePrimeMismatch,
+			enclosedThreePrimeMismatch, fc.energyParams)
+
+		// seek to the next stem
+		enclosedFivePrimeIdx = enclosedThreePrimeIdx + 1
+
+		for enclosedFivePrimeIdx < closingThreePrimeIdx && pairTable[enclosedFivePrimeIdx] == -1 {
+			enclosedFivePrimeIdx++
+		}
+
+		// add unpaired nucleotides
+		nbUnpairedNucleotides += enclosedFivePrimeIdx - enclosedThreePrimeIdx - 1
+	}
+
+	if closingFivePrimeIdx > 0 {
+		// actual closing pair
+
+		// vivek: Is this a bug? Why are the five and three prime mismatches opposite?
+		// Created an issue in the original repo: https://github.com/ViennaRNA/ViennaRNA/issues/126
+		basePairType := encodedBasePairType(fc.sequence[closingThreePrimeIdx],
+			fc.sequence[closingFivePrimeIdx], fc.energyParams.basePairEncodedTypeMap)
+		closingFivePrimeMismatch := fc.encodedSequence[closingThreePrimeIdx-1]
+		closingThreePrimeMismatch := fc.encodedSequence[closingFivePrimeIdx+1]
+
+		energy += multiLoopStemEnergy(basePairType, closingFivePrimeMismatch,
+			closingThreePrimeMismatch, fc.energyParams)
+	} else {
+		/* virtual closing pair */
+		energy += multiLoopStemEnergy(0, -1, -1, fc.energyParams)
+	}
+
+	// add bonus energies for unpaired nucleotides
+	energy += nbUnpairedNucleotides * fc.energyParams.MLbase
+
+	energyContribution := EnergyContribution{
+		closingFivePrimeIdx:  closingFivePrimeIdx,
+		closingThreePrimeIdx: closingThreePrimeIdx,
+		energy:               energy,
+		loopType:             MultiLoop,
+	}
+	return energy, energyContribution
+}
+
+/**
+ *  Compute the energy contribution of a multi-loop stem.
+ *
+ *  Given a base pair (i,j) (encoded by `basePairType`), compute the
+ *  energy contribution including dangling-end/terminal-mismatch contributions.
+ *  You can prevent taking 5'-, 3'-dangles or mismatch contributions into
+ *  account by passing -1 for `fivePrimeMismatch` and/or `threePrimeMismatch`
+ *  respectively.
+ *
```

Comment:
How hard would it be to make a diagram like those above? They were very helpful.